compiler/src/dotty/tools/dotc/transform/init/Checking.scala

```diff
@@ -254,34 +254,98 @@ object Checking {
         val Summary(pots, effs) = expand(pot)
         val effs2 = pots.map(FieldAccess(_, field)(eff.source))
         (effs2 ++ effs).toList.flatMap(check(_))
+    }
+
+  /// Check if we can just directly promote a potential.
+  /// A potential can be (currently) directly promoted if and only if:
+  /// - `pot == this` and all fields of this are initialized, or
+  /// - `pot == Warm(C, outer)` where `outer` can be directly promoted.
+  private def canDirectlyPromote(pot: Potential)(using state: State): Boolean = trace("checking direct promotion of " + pot.show, init) {
+    if (state.safePromoted.contains(pot)) true
+    else pot match {
+      case pot: ThisRef =>
+        // If we have all fields initialized, then we can promote This to hot.
+        val classRef = state.thisClass.info.asInstanceOf[ClassInfo].appliedRef
+        classRef.fields.forall { denot =>
+          val sym = denot.symbol
+          sym.isOneOf(Flags.Lazy | Flags.Deferred) || state.fieldsInited.contains(sym)
+        }
+      case Warm(cls, outer) =>
+        canDirectlyPromote(outer)
+      case _ =>
+        val summary = expand(pot)
```

Comment:
This will lead to infinite loops. The following might be a test case:

```Scala
class A {
  val a = f
  def f: A = B
  class B(x: Int)

  println(new a.B(5))
  val n = 10
}
```

Reply:
Mitigating this with a `visited` set. I think it's easier to just set a maximum recursion depth though, because this is only a speed-up measure most of the time...

Reply:
Yes, you can play with it and see which works better for the tests we have.